src/Strategy/Python/Poetry.hs

```diff
@@ -0,0 +1,187 @@
+getDeps project = do
+  context "Poetry" $ context "Static analysis" $ analyze project
+
+-- | Analyzes Poetry Project and creates dependency graph.
+analyze ::
+  ( Has ReadFS sig m
+  , Has Diagnostics sig m
+  , Has Logger sig m
+  ) =>
+  PoetryProject ->
+  m (Graphing Dependency)
+analyze PoetryProject{pyProjectToml, poetryLock} = do
+  pyproject <- readContentsToml pyProjectCodec (pyProjectTomlPath pyProjectToml)
+  case poetryLock of
+    Just lockPath -> do
+      poetryLockProject <- readContentsToml poetryLockCodec (poetryLockPath lockPath)
+      _ <- logIgnoredDeps pyproject (Just poetryLockProject)
+      context "Building dependency graph from pyproject.toml and poetry.lock" $ pure $ buildGraphWithLock poetryLockProject pyproject
+    Nothing -> context "Building dependency graph from only pyproject.toml" $ pure $ Graphing.fromList $ pyProjectDeps pyproject
+
+-- | Build graph of `Dependency` only from `PyProject` and `PoetryLock`.
+buildGraphWithLock :: PoetryLock -> PyProject -> Graphing Dependency
+buildGraphWithLock lockProject poetryProject = promoteToDirect isDirect graph
+  where
+    -- Dependencies in `poetry.lock` are direct if they're specified in `pyproject.toml`.
+    isDirect :: Dependency -> Bool
+    isDirect dep = case pyprojectPoetry poetryProject of
+      Nothing -> False
+      Just _ -> any (\n -> dependencyName n == dependencyName dep) (pyProjectDeps poetryProject)
+
+    lowerCasedPkgName :: PackageName -> PackageName
+    lowerCasedPkgName name = PackageName . toLower $ unPackageName name
+
+    graph = gmap pkgNameToDependency (buildPackageNameGraph $ poetryLockPackages lockProject)
+    mapOfDependency = toMap $ poetryLockPackages lockProject
+
+    -- Pip packages are [case insensitive](https://www.python.org/dev/peps/pep-0508/#id21), but poetry.lock may use
+    -- non-canonical name for reference. Try to lookup with provided casing, otherwise fallback to lower casing.
+    --
+    -- Poetry.lock when referencing deep dependencies uses non canonical names.
+    --
+    --  ```toml
+    --  [package.dependencies]
+    --  MarkupSafe = ">=2.0"
+    --  ....
+    --
+    -- [[package]]
+    -- name = "markupsafe"
+    -- version = "2.0.1"
+    -- ...
+    -- ```
+    --
+    -- TODO: Better approach to handle casing scenarios with poetry.lock file.
+    pkgNameToDependency name =
+      fromMaybe
+        ( Dependency
+            { dependencyType = PipType
+            , dependencyName = unPackageName name
+            , dependencyVersion = Nothing
+            , dependencyLocations = []
+            , dependencyEnvironments = []
+            , dependencyTags = Map.empty
+            }
+        )
+        $ Map.lookup name mapOfDependency
+          <|> Map.lookup (lowerCasedPkgName name) mapOfDependency
+
+-- | Builds the Package Name Graph.
+-- In python, package names are unique, and only single version of a package can be used.
```

Comment:
```suggestion
-- | Using a Poetry lockfile, build the graph of packages.
-- The resulting graph contains edges, but does not distinguish between direct and deep dependencies, since `poetry.lock` does not indicate which dependencies are direct.
```